### PR TITLE
poetry-plugin-up: init at 0.2.1

### PR DESCRIPTION
**Pull request by K900**
###### Description of changes

A plugin to update dependencies to latest even if pyproject.toml says not to.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/package-management/poetry-plugin-up/default.nix [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+, fetchFromGitHub
+, python3
+}:
+
+let
+  python = python3;
+in
+python.pkgs.buildPythonPackage rec {
+  pname = "poetry-plugin-up";
+  version = "0.2.1";
```
Comment (mweinelt) [inline pkgs/tools/package-management/poetry-plugin-up/default.nix:11]: format after version, please

File: pkgs/tools/package-management/poetry-plugin-up/default.nix [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+  src = fetchFromGitHub {
+    owner = "MousaZeidBaker";
+    repo = pname;
+    rev = "refs/tags/${version}";
+    hash = "sha256-16p0emvgWa56Km8U5HualCSStbulqyINbC3Jez9Y1n0=";
+  };
+
+  format = "pyproject";
+  nativeBuildInputs = with python.pkgs; [ poetry-core pythonRelaxDepsHook ];
```
Comment (mweinelt) [inline pkgs/tools/package-management/poetry-plugin-up/default.nix:21]:
```suggestion
  nativeBuildInputs = with python.pkgs; [
    poetry-core
    pythonRelaxDepsHook
  ];

```

one dependency per line and a free line between the lists.

File: pkgs/tools/package-management/poetry-plugin-up/default.nix [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  # break dep cycle
+  pythonRemoveDeps = [ "poetry" ];
```
Comment (mweinelt) [inline pkgs/tools/package-management/poetry-plugin-up/default.nix:23]:
```suggestion
  pythonRemoveDeps = 
    # break dependency cycle
    "poetry"
  ];
```

File: pkgs/tools/package-management/poetry-plugin-up/default.nix [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+  meta = with lib; {
```
Comment (mweinelt) [inline pkgs/tools/package-management/poetry-plugin-up/default.nix:25]:
```suggestion
  meta = with lib; {
    changelog = "https://github.com/MousaZeidBaker/poetry-plugin-up/releases/tag/${version}";
```

**Review by mweinelt — CHANGES_REQUESTED** [review]
Use `pytestCheckHook` to run the tests.

File: pkgs/top-level/all-packages.nix [hunk 5]
```diff
@@ -16383,6 +16383,8 @@ with pkgs;
 
   poetry = callPackage ../tools/package-management/poetry { };
 
+  poetry-plugin-up = callPackage ../tools/package-management/poetry-plugin-up { };
```
Comment (dotlambda) [inline pkgs/top-level/all-packages.nix:16386]: This should be in `python-packages.nix`.

Comment (K900) [reply pkgs/top-level/all-packages.nix:16386]: Why? It's not going to be useful with any Python version but the one Poetry is built with.

Comment (mweinelt) [reply pkgs/top-level/all-packages.nix:16386]: It could also live at `poetry.plugins` or something like that.

**Comment by K900** [comment]
> Use `pytestCheckHook` to run the tests.

Tests won't run because we don't have a Poetry.

File: pkgs/top-level/all-packages.nix (hunk 5 above)
Comment (K900) [reply pkgs/top-level/all-packages.nix:16386]: I thought about that but I'm not sure if it's worth adding another scope when we only have one package to put under it.

**Comment by K900** [comment]
Applied all the style fixes.

**Comment by mweinelt** [comment]
> Tests won't run because we don't have a Poetry.

Add poetry to checkInputs, disable `doCheck`, import the package itself, create an override in passthru.tests á la `poetry-plugin-up.overridePythonAttrs (_: { doCheck = true; })`

**Comment by K900** [comment]
That's still not going to work because Poetry doesn't propagate its dependencies (see #209534).

File: pkgs/top-level/all-packages.nix (hunk 5 above)
Comment (dotlambda) [reply pkgs/top-level/all-packages.nix:16386]: It is. We can ensure that plugins use the same Python package set in two ways: the `plugins` argument becomes a function that is passed `python.pkgs` (see home-assistant) and plugins are packaged in `python3Packages`, or plugins live in `poetry.plugins` where they're automatically built with the correct Python package set.

File: pkgs/tools/package-management/poetry/default.nix [hunk 6]
```diff
@@ -3,122 +3,158 @@
...
+      trove-classifiers
+      virtualenv
+    ] ++ lib.optionals (stdenv.isDarwin) [
+      xattr
+    ] ++ lib.optionals (pythonOlder "3.11") [
+      tomli
+    ] ++ lib.optionals (pythonOlder "3.10") [
+      importlib-metadata
+    ] ++ lib.optionals (pythonOlder "3.8") [
+      backports-cached-property
```
Comment (SuperSandro2000) [inline pkgs/tools/package-management/poetry/default.nix:54]: We can remove that, we no longer support 3.8

**Comment by K900** [comment]
Rebased to catch the nativeCheckInputs change.

**Review by dotlambda — COMMENTED** [review]
I think this is more complicated than it needs to be. I don't see why one type of wrapping shouldn't suffice and I'm not fond of the `$PYTHONPATH` hackery.
Thus I tried to achieve the same thing just using `overridePythonAttrs` and thought that it's maybe a good idea to put the actual expression for poetry in a separate file `unwrapped.nix`. See https://github.com/NixOS/nixpkgs/pull/213914 for my suggestion.

**Comment by K900** [comment]
The reason I went with a double wrapper here is mostly to avoid rebuilding poetry itself, which is not _particularly_ slow but also kinda slow.

**Comment by dotlambda** [comment]
> The reason I went with a double wrapper here is mostly to avoid rebuilding poetry itself, which is not _particularly_ slow but also kinda slow.

That does make sense indeed. I guess I could copy the double wrapper in my approach if that's preferred.
I still think having `poetry.plugins` is a bit cleaner and makes sure all plugins use the same versions of all Python packages.

**Comment by K900** [comment]
Cleaner than what? Sorry, might be missing something...

**Comment by dotlambda** [comment]
> Cleaner than what? Sorry, might be missing something...

You're right, your approach is almost the same in this regard. So never mind.

File: pkgs/tools/package-management/poetry/default.nix [hunk 7]
```diff
@@ -18,118 +19,153 @@ let
...
-}
+
+  wrapPoetry = { plugins ? [] }: let
+    wrapNoPlugins = ''
+      ln -s ${unwrapped}/bin/poetry $out/bin/poetry
+    '';
+
+    wrapWithPlugins = ''
+      buildPythonPath $plugins
+      makeWrapper ${unwrapped}/bin/poetry $out/bin/poetry --suffix PYTHONPATH : "$program_PYTHONPATH"
```
Comment (dotlambda) [inline pkgs/tools/package-management/poetry/default.nix:129]: This is basically a no-op if `$program_PYTHONPATH` is empty. Is there a specific reason you're using a different wrapper in that case?

Comment (K900) [reply pkgs/tools/package-management/poetry/default.nix:129]: It's a tiny efficiency thing, probably not worth worrying about realistically.

**Comment by mweinelt** [comment]
#213914 looks simpler and is the more python-native approach using `packageOverrides` for the package set and `overridePythonAttrs` to propagate plugins into the poetry package. 

I think I prefer it also because it relies less on a bespoke wrapping logic, like is the case here.

The downside of the override approach is that you get to rebuild poetry, which takes a minute on my desktop due to five very slow tests. We could however also just disable tests on the overridden derivation to get that penalty out of the way.

**Comment by K900** [comment]
I definitely like the `packageOverrides` approach and probably would have done the same thing if I remembered it existed at the time (lol). I'm still not 100% convinced on the wrapper versus adding inputs part, but it's also definitely not a blocker and either way should be fine.

**Comment by dotlambda** [comment]
> We could however also just disable tests on the overridden derivation to get that penalty out of the way.

That's a good compromise.

> I'm still not 100% convinced on the wrapper versus adding inputs part

The advantage of adding inputs is that it's going to be correct in the future even if `wrapPython`'s logic changes. We also won't have to change the wrapper if files are addded to `poetry`'s `$out/share` or whatever.

**Comment by K900** [comment]
Then I'm just going to close this in favor of https://github.com/NixOS/nixpkgs/pull/213914.

**Pull request closed without merging**